### PR TITLE
Added Webhooks support for docket alerts

### PR DESCRIPTION
**Pull request by albertisfu**
The Webhook page was updated according to the Figma proposal, some screenshots are below:

**Webhooks page, no webhooks registered.**
Screen Shot 2022-07-11 at 10 01 06

**Add Webhook modal:**
According to the requirements.  Right now it's possible for a user to have only one webhook for each type.
When adding a new Webhook, the Event Type field will only show the users' remaining event types options.
Screen Shot 2022-07-11 at 10 02 36

When a user tries to add a new Webhook after adding one webhook of each type, a message will be shown.
Screen Shot 2022-07-11 at 10 04 46

When updating a Webhook it's only possible to change the URL and enabled status. To avoid conflicts due to duplicating event types.
Screen Shot 2022-07-11 at 10 03 40

When deleting a Webhook a confirmation dialog is shown:
Screen Shot 2022-07-11 at 10 05 39

Some additional comments:

- As described above, the one type webhook constraint for each user works at the interface level.  Should we also add a DB constraint? I didn't add it due to the comment that it might be needed to add support for multiple webhooks of each type in the future.

- According to requirements, Webhooks for DocketAlerts and Recap.email are now the same. Some fixes and changes were applied to accomplish it. The send_docket_to_webhook task was moved into alerts/tasks to avoid a circular reference.

- Webhooks for DocketAlerts are sent to users that have an enabled DOCKET_ALERT Webhook. The webhook is triggered only for cases to which the user is subscribed. For recap.email subscriptions, the webhook is triggered the first time only if the user has the "auto-subscribe" option enabled, took advantage of the existing logic to avoid doing additional queries to obtain webhook recipients.

- One question that popped up, is it necessary to retry the `send_docket_to_webhook` task on a RequestException or Timeout? Since we are going to count the webhook failures.

- I'll add support for Search Alerts and Recap Fetch webhooks in the next PRs.

Please let me know if additional changes are needed over the interface or logic.

### CODE REVIEW

File: cl/users/urls.py [hunk 1]
```diff
@@ -102,6 +102,21 @@
...
+    ),
+    path(
+        "profile/webhooks-list/",
+        views.view_webhooks_list,
+        name="view_webhooks_list",
+    ),
+    path(
+        "profile/webhook-delete/<int:pk>/",
+        views.delete_webhook,
+        name="delete_webhook",
```
Comment (mlissner) [inline cl/users/urls.py:118]: I just read many tutorials on HTMX and Django, and I don't like any of them. They all take the approach you did here, which uses multiple views separating out logic. Doing this leaves our views untethered to each other and spreads our code out very thinly. 

I think a better way to do this is to do as few views as possible, and to use HTTP verbs. So:

 - HTTP DELETE sent to /api/rest/v3/webhooks/<id>/ deletes something
 - HTTP PUT sent to /api/rest/v3/webhooks/<id>/ updates something
 - HTTP POST sent to /api/rest/v3/webhooks/ creates something
 - HTTP GET sent to /api/rest/v3/webhooks/ lists them
 - HTTP GET sent to /api/rest/v3/webhooks/<id> lists one (if needed)

This mostly mirrors the approach in Django REST Framework, and you can see that I used its URLs in my above concept. 

I spent some time looking for an integration between DRF and HTMX, but didn't find anything good. It seems like it should exist, but I guess not (yet). 

In any case, I think we should do something like the above. If we can use DRF to power HTMX, we should do that, using the URLs above. If we can't, we should use HTTP verbs and as few URL paths as possible, probably putting them at `/webhooks/` instead of under `/api/rest/v3/`.

Does this make sense? Do you agree?

Comment (mlissner) [reply cl/users/urls.py:118]: The other thing that might be worth doing is to use django-htmx to combine views. For example, you have two that basically just list webhooks. Those could be combined by using the [django-htmx middleware](https://django-htmx.readthedocs.io/en/latest/middleware.html#django_htmx.middleware.HtmxDetails.__bool__).

File: cl/users/views.py [hunk 2]
```diff
@@ -712,20 +713,69 @@ def moosend_webhook(request: HttpRequest) -> HttpResponse:
...
+
+@login_required
+@require_http_methods(["POST", "GET"])
+def webhook_form(request: HttpRequest) -> HttpResponse:
+    """Handle the creation or editing of a webhook"""
+    pk = ""
+    if "pk" in request.GET:
+        pk = request.GET["pk"]
+    if "pk" in request.POST:
+        pk = request.POST["pk"]
```
Comment (mlissner) [inline cl/users/views.py:756]: Maybe this comment won't matter if we redo this, but why do we allow a GET request here?

File: cl/users/templates/includes/webhooks-modal.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+     class="modal hidden-print text-left"
+     role="dialog"
+     aria-hidden="true"
+     tabindex="-1">
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span>
+        </button>
+        <h2 class="modal-title">Add a Webhook endpoint</h2>
```
Comment (mlissner) [inline cl/users/templates/includes/webhooks-modal.html:12]: You capitalized webhook a lot in here. Let's keep them uncapitalized most of the time:

```suggestion
        <h2 class="modal-title">Add a webhook endpoint</h2>
```

File: cl/users/templates/profile/webhooks.html [hunk 4]
```diff
@@ -1,68 +1,96 @@
...
+{% load static %}
 {% load waffle_tags %}
 
 {% block title %} Developer Tools &ndash; CourtListener.com{% endblock %}
 
 {% block nav-api %}active{% endblock %}
+{% block footer-scripts %}
+  {% if DEBUG %}
+    <script src="{% static "js/htmx.js" %}"></script>
+    <link href="{% static "css/bootstrap4-toggle.css" %}" rel="stylesheet">
```
Comment (mlissner) [inline cl/users/templates/profile/webhooks.html:11]: Did you end up using this?

File: cl/users/templates/profile/webhooks.html [hunk 5]
```diff
@@ -1,68 +1,96 @@
...
+      let webhook_form = document.getElementById("webhooks-body");
+      if (e.detail.target.id == "webhooks-body") {
+        //Change submit button text according to the action performed
+        if (e.detail.requestConfig.parameters.hasOwnProperty('update')) {
+          webhook_submit.innerHTML = "Update Webhook";
+        }
+        // If user has reached the max number of webhooks, show a message.
+        let event_type_options = document.getElementById("id_event_type").options.length;
+        if(event_type_options == 0){
+          webhook_form.innerHTML = "<b class='text-center'>You've reached the maximum number of webhooks. Please delete one before adding another.</b>";
```
Comment (mlissner) [inline cl/users/templates/profile/webhooks.html:30]:
```suggestion
          webhook_form.innerHTML = "<b class='text-center'>You already have a webhook configured for each type of event. Please delete one before making another.</b>";
```

File: cl/users/templates/profile/webhooks.html [hunk 6]
```diff
@@ -1,68 +1,96 @@
...
+                  </tbody>
+                </table>
               </div>
             </div>
-          </form>
+          </div>
         </div>
       </div>
+      {% include "includes/webhooks-modal.html" %}
+      <p>Webhooks let you send a signal to your server whenever certain events happen on our website. To create a webhook, use the form below.</p>
```
Comment (mlissner) [inline cl/users/templates/profile/webhooks.html:91]:
```suggestion
      <p>Webhooks enable bi-directional APIs by sending data to your server whenever events happen on ours. Use webhooks to get the fastest updates &mdash; without polling.</p>
```

File: cl/users/templates/includes/webhooks-list.html [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+          <span class="text-success">Active</span>
+        {% else %}
+          <span class="text-danger">Disabled</span>
+        {% endif %}
+      </p>
+    </td>
+    <td>{{ webhook.url }}</td>
+    <td class="right" hx-headers='{"X-CSRFToken": "{{ csrf_token }}"}'>
+      <button hx-post="{% url "delete_webhook" webhook.pk %}"
+              hx-confirm="Are you sure you want to delete this Webhook?"
```
Comment (mlissner) [inline cl/users/templates/includes/webhooks-list.html:21]:
```suggestion
              hx-confirm="Are you sure you want to delete this webhook?"
```

File: cl/recap/tasks.py [hunk 8]
```diff
@@ -1658,11 +1606,8 @@ def process_recap_email(
             recap_email_recipients = get_recap_email_recipients(
                 data["email_recipients"]
             )
-            chain(
-                send_docket_alert.si(
-                    docket.pk, start_time, recap_email_recipients
-                ),
-                send_docket_to_webhook.si(docket.pk, start_time, epq.pk),
+            send_docket_alert.si(
+                docket.pk, start_time, recap_email_recipients
```
Comment (mlissner) [inline cl/recap/tasks.py:1610]: If we're not doing a chain and don't need special stuff provided by apply_async, we should just use .delay().

File: cl/alerts/tasks.py [hunk 9]
```diff
@@ -163,7 +222,7 @@ def send_docket_alert(
         return
 
     # Notify every user that's subscribed to this alert.
-    messages = make_alert_messages(
+    messages, da_webhook_recipients = make_alert_messages(
```
Comment (mlissner) [inline cl/alerts/tasks.py:225]: We need refactoring here. This is the general stuff we do:

 - Get recipients for messages
 - Get recipients for webhooks
 - Generate emails
 - Generate webhooks
 - Send emails
 - Send webhooks

Right now we call send_docket_alerts, which calls make_alert_messages and send_docket_to_webhook. make_alert_messages, makes the messages, and gathers a list of webhook recipients, and sends messages. send_docket_to_webhook takes recipients from make_alert_messages, then generates webhooks and sends them. This is all very confusing! 

Can you see if you can simplify this so that, perhaps, it more closely mirrors the six steps above? Somebody calls something like send_alerts_and_webhooks, and then it gets recipients for each, generates content for each, and sends each?

**Review by mlissner — CHANGES_REQUESTED** [review]
Thanks for this, Alberto. I think this looks great and it works well, but I want some tweaks to the code that'll make our lives easier and better in the future. The two big ones are:

1. I think we should be using more restful patterns with HTMX. I'm surprised nobody is, but I wrote more about how I think it should work in a comment.

2. I think we've reached the point where the alert and webhook code has become hard to follow and maintain, and it needs a refactor. Details in a separate comment.

On top of those big ones, a few small ones:

1. We need `noop` on the migration.
4. The enabled box in the modal should be aligned with the text to the left. Bootstrap usually makes this easy (or I saw you wanted to use a toggle for this instead, but I think the toggle is probably the wrong design here).
5. We need to write webhook documentation for that link. Maybe do this as a follow-up issue? Do we have an issue for that? Do you want to make one?
6. We're generating webhook event logs, but not showing them. Do you have something in mind for how to show them? 

Thank you.

File: cl/users/urls.py (hunk 1 above)
Comment (albertisfu) [reply cl/users/urls.py]: Thanks! this is a better approach to implementing htmx with django, I managed to use DRF to render over an HTML template (needed for htmx) using a ModelViewSet I could implement all the necessary actions for creating, editing, updating, and deleting webhooks.

Now it's possible to handle all the actions using one route: `/api/rest/v3/webhooks` and different verbs such as DELETE, PUT, POST, and GET.

In order to use `/api/rest/v3/webhooks` added this router on `cl.api.urls `, however, there is something that makes me think maybe that's wrong.  Because if you go to `/api/rest/v3/` it's possible to see the webhooks endpoint but this is not a normal API endpoint due to its actions render to an HTML template as default, would it be better to move it out of `/api/rest/v3/`?

File: cl/users/templates/includes/webhooks-modal.html (hunk 3 above)
Comment (albertisfu) [reply cl/users/templates/includes/webhooks-modal.html]: changed!

File: cl/users/templates/profile/webhooks.html (hunk 4 above)
Comment (albertisfu) [reply cl/users/templates/profile/webhooks.html]: no, we don't use the toggle here, removed.

File: cl/users/templates/profile/webhooks.html (hunk 5 above)
Comment (albertisfu) [reply cl/users/templates/profile/webhooks.html]: changed!

File: cl/users/templates/profile/webhooks.html (hunk 6 above)
Comment (albertisfu) [reply cl/users/templates/profile/webhooks.html]: changed!

File: cl/users/templates/includes/webhooks-list.html (hunk 7 above)
Comment (albertisfu) [reply cl/users/templates/includes/webhooks-list.html:21]: changed.

File: cl/recap/tasks.py (hunk 8 above)
Comment (albertisfu) [reply cl/recap/tasks.py:1610]: changed!

**Comment by albertisfu** [comment]
Thank you for your comments and suggestions!

I still continue working on the refactor of send docket alerts and webhooks methods but I wanted to submit these changes and comments to hear your opinion.

- I've applied the refactor of webhook views to use a Restful approach with HTMX. It was possible to use DRF for it, there is a comment about this above.
- I've opened the issue to write webhook documentation here #2176
- About Logs, yeah I think we could add a section to show logs to users, I added a proposal on [Figma](https://www.figma.com/file/l3hda5VzQw6QZyKckqlne9/Webhooks?node-id=0%3A1) for it with some comments, some screenshots below.
- I also was thinking that it would be helpful for users to add a feature that users can use to test its webhook integration. It might be something simple as allowing users to trigger an example webhook request for each event type, what do you think?
- I'll made some research about webhooks retries to make a proposal of retry back off, I have a couple of ideas on mind but I just want to be sure how other services handle webhooks retries.



Screen Shot 2022-07-15 at 12 47 26
Screen Shot 2022-07-15 at 12 47 41

File: cl/recap/tasks.py (hunk 8 above)
Comment (mlissner) [reply cl/recap/tasks.py:1610]: There's an even better way to do this. Check out the example here, which drops the `si` too: https://docs.celeryq.dev/en/stable/userguide/calling.html#basics

File: cl/alerts/tasks.py (hunk 9 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: I've refactored these methods as suggested, I hope with these changes everything is more clear.

- Now somebody calls `send_alert_and_webhook` or `send_alerts_and_webhooks` to start the sending docket alerts and webhooks process.
- `send_alert_and_webhook` first calls `get_docket_alert_recipients` to get recipients for docket alerts.
- `get_docket_alert_recipients` returns recipients for each type of notification including the new admin's notification to inform about @recap.email users that don't exist.
- Then we call `send_recap_email_user_not_found` to send the notification to admins if there are any @recap.email users that don't exist.
- Then we compose each type of docket alert message:
  - `docket_alert_messages`: messages for users that are currently subscribed to the case
  - `re_first_time_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option enabled.
  - `re_first_time_no_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option disabled.
- Then we join the different types of messages and sent them at once.
- Finally, we get webhooks recipients and sent them.

File: cl/recap/tasks.py (hunk 8 above)
Comment (albertisfu) [reply cl/recap/tasks.py]: Thanks, I've changed it.

**Comment by albertisfu** [comment]
I've submitted additional changes that refactor send docket alerts and webhooks tasks. I hope now it's simpler to follow the process, more details about this refactoring are in a comment above.

Let me know what you think.

Thanks

File: cl/users/urls.py (hunk 1 above)
Comment (mlissner) [reply cl/users/urls.py]: Yes, I agree, now that I see it in /api/rest/v3/webhooks, it feels wrong. Let's give it a better home. A few possibilities:

 - /api/htmx/webhooks/  - I kind of like this.
 - /profile/webhooks/ - Feels limited, but maybe it's fine?
 - /webhooks/ - This feels like it's too high in the hierarchy.

What do you think?

**Comment by mlissner** [comment]
> I also was thinking that it would be helpful for users to add a feature that users can use to test its webhook integration. It might be something simple as allowing users to trigger an example webhook request for each event type, what do you think?

Yes, we should make an issue for this and put it on our backlog.

File: cl/alerts/tasks.py [hunk 10]
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
...
     return messages
 
 
 # Ignore the result or else we'll use a lot of memory.
 @app.task(ignore_result=True)
-def send_docket_alert(
+def send_alert_and_webhook(
     d_pk: int,
     since: datetime,
     recap_email_recipients: list[str] = [],
```
Comment (mlissner) [inline cl/alerts/tasks.py:166]: I guess I missed this before, but `[]` is mutable and will cause all kinds of fun problems, probably. Instead this should be `= None`, and then you should add a something like:

```
if recap_email_recipients is None:
    recap_email_recipients = []
```

File: cl/alerts/tasks.py [hunk 11]
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
...
 
 
+def get_docket_alert_recipients(
+    d_pk: int,
+    recap_email_recipients: list[str] = [],
+) -> tuple[
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[str],
```
Comment (mlissner) [inline cl/alerts/tasks.py:40]: OK, this refactor isn't quite it, and this return type should have been the "[code-smell](https://en.wikipedia.org/wiki/Code_smell)" that made you realize something was wrong. 

Linus Torvalds once said something like, "when you've got your data objects sorted out and as simple as possible, you know you're done." This ain't that. 

The solution here is hinted at by the name of the function. We want to get recipients, so it's going to be a list. But a list of what? We could use dicts like:

```
email: 'mike@free.law',
secret_key: 'asdfasdfasdfa',
first_time: True,
auto_subscribe: False,
does_not_exist: False,
```

Or, dare I say it: We could use an object to hold that information (this would make typing easier).

Then the return from this function would be something like `list(MessageData)` (or whatever). This is all notional, meaning I want you to make it good, but this return type is a major code smell.

File: cl/alerts/tasks.py [hunk 12]
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
...
+    recap_email_first_time_no_auto_subscribe_list = []
+    # List of @recap.email addresses that don't belong to any user
+    recap_email_user_does_not_exist_list = []
+
+    # First, get current docket alert recipients to avoid duplicate alerts
+    docket_alerts_current_subscribers = DocketAlert.objects.filter(
+        docket_id=d_pk, alert_type=DocketAlert.SUBSCRIPTION
+    )
+    for da in docket_alerts_current_subscribers:
+        recipient_tuple = (da.user.email, da.secret_key)
```
Comment (mlissner) [inline cl/alerts/tasks.py:62]: We need to remember to focus on queries in this code. So if you need to get `da.user.email`, remember to use a `select_related` up above.

File: cl/alerts/tasks.py [hunk 13]
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
...
+            recap_email_first_time_auto_subscribe_list.append(recipient_tuple)
+        else:
+            # First time recipient, auto_subscribe False
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk,
+                user=user_profile.user,
+                alert_type=DocketAlert.UNSUBSCRIPTION,
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_no_auto_subscribe_list.append(
```
Comment (mlissner) [inline cl/alerts/tasks.py:95]: This `objects.create(` code is almost completely duplicated. I think when you make the refactor to return a list of dicts/objects, this should simplify. It should wind up with something like:

```
    # Somewhere above: md = MessageData()
    alert_type = DocketAlert.SUBSCRIPTION if auto_subscribe else DocketAlert.UNSUBSCRIPTION
    docket_alert = DocketAlert.objects.create(
        docket_id=d_pk,
        user=user_profile.user,
        alert_type=alert_type,
    )
    md.email_address = email_address
    md.secret_key = docket_alert.secret_key
    md.auto_subscribe = user_profile.auto_subscribe
    message.append(md)
return messages
```

File: cl/alerts/tasks.py (hunk 11 above)
Comment (mlissner) [reply cl/alerts/tasks.py:40]: Oh, I guess a data class is probably a good tool here too: https://docs.python.org/3/library/dataclasses.html

File: cl/alerts/tasks.py [hunk 14]
```diff
@@ -168,25 +200,58 @@ def send_docket_alert(
...
+        recap_email_no_auto_recipients,
+        first_email=True,
+        auto_subscribe=False,
+    )
+
+    messages = (
+        docket_alert_messages
+        + re_first_time_auto_subscribe_messages
+        + re_first_time_no_auto_subscribe_messages
     )
```
Comment (mlissner) [inline cl/alerts/tasks.py:229]: After the refactor I mention above, I think this will all drop away. Aaaaaahhhh.

File: cl/alerts/tasks.py [hunk 15]
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
...
     """
 
     case_name = trunc(best_case_name(d), 100, ellipsis="...")
-    messages = []
     txt_template = loader.get_template("docket_alert_email.txt")
     html_template = loader.get_template("docket_alert_email.html")
     subject_template = loader.get_template("docket_alert_subject.txt")
     subject_context = {
         "docket": d,
         "count": new_des.count(),
```
Comment (mlissner) [inline cl/alerts/tasks.py:130]: You'll save a query by making this `len(new_des)`. Just saying!

File: cl/alerts/tasks.py [hunk 16]
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
...
+def send_docket_to_webhook(
+    d_pk: int,
+    since: datetime,
+    webhook_recipients: list[str],
+) -> None:
+    """POSTS the DocketAlert to the recipients webhook(s)
+
+    :param d_pk: The Docket primary key
+    :param since: Start time for querying the docket entries
+    :webhook_recipients: The list of email recipients to send the webhook to
```
Comment (mlissner) [inline cl/alerts/tasks.py:324]:
```suggestion
    :param webhook_recipients: The list of email recipients to send the webhook to
```

File: cl/alerts/tasks.py [hunk 17]
```diff
@@ -168,25 +200,58 @@ def send_docket_alert(
...
+
+    # Webhooks recipients are handled separately, send it to current docket
+    # alerts recipients + first time recap.email recipients with auto_subscribe
+    # set to true.
+    webhook_recipients = (
+        docket_alert_recipients + recap_email_auto_subscribe_recipients
+    )
+    # Send the docket to webhook
+    webhook_email_recipients = [email[0] for email in webhook_recipients]
+    send_docket_to_webhook.delay(d_pk, since, webhook_email_recipients)
```
Comment (mlissner) [inline cl/alerts/tasks.py:245]: Passing the email addresses down to the webhook is an interesting optimization, but I think it's not worth the confusion ("Why are we sending email addresses to the webhook?") or the tethering ("Why do we need to send alerts before we can send webhooks?").

Also, the name of the function is a bit odd. Shouldn't it be `send_docket_alert_webhooks`?

So, with those two changes, I think the signature here should change to:
```
def send_docket_alert_webhooks(d_pk: int, since: datetime):
```

File: cl/alerts/tasks.py [hunk 18]
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
...
+        date_created__gte=since, docket_id=d_pk
+    )
+    if docket_entries.count() == 0:
+        # No new docket entries.
+        return
+
+    webhooks = Webhook.objects.filter(
+        event_type=WebhookEventType.DOCKET_ALERT,
+        user__email__in=webhook_recipients,
+        enabled=True,
```
Comment (mlissner) [inline cl/alerts/tasks.py:338]: In my last comment I said not to pass down the webhook_recipients variable, so now we need to tweak this query to get "Enabled webhooks of type DOCKET_ALERT where the user is subscribed to d_pk."

I admit I gave this a bit of thought and couldn't get it quite figured out in my head, but I'm sure you'll nail it.

File: cl/alerts/tasks.py [hunk 19]
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
...
+    serialized_docket_entries = []
+    for de in docket_entries:
+        serialized_docket_entries.append(DocketEntrySerializer(de).data)
+
+    for webhook in webhooks:
+        post_content = {
+            "webhook": {
+                "event_type": webhook.event_type,
+                "version": webhook.version,
+                "date_created": webhook.date_created.isoformat(),
```
Comment (mlissner) [inline cl/alerts/tasks.py:350]: Don't forget to do the `deprecation_date` field (or similar) here.

File: cl/alerts/tasks.py [hunk 20]
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
...
+        WebhookEvent.objects.create(
+            webhook=webhook,
+            status_code=response.status_code,
+            content=post_content,
+            response=response.text,
+        )
+
+        if not response.ok:
+            webhook.failure_count = webhook.failure_count + 1
+            webhook.save()
```
Comment (mlissner) [inline cl/alerts/tasks.py:364]: We can avoid this query by checking response.ok before the `create`, right?

File: cl/alerts/tasks.py [hunk 21]
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
...
+
+    webhooks = Webhook.objects.filter(
+        event_type=WebhookEventType.DOCKET_ALERT,
+        user__email__in=webhook_recipients,
+        enabled=True,
+    )
+
+    serialized_docket_entries = []
+    for de in docket_entries:
+        serialized_docket_entries.append(DocketEntrySerializer(de).data)
```
Comment (mlissner) [inline cl/alerts/tasks.py:343]: I'm a bit surprised you can get away with such a simple serialization. Check out the hoops we have to go through for the bulk data serializer. https://github.com/freelawproject/courtlistener/blob/0c74a05a74188066f268c7023f2a12b0f583cb8e/cl/api/tasks.py#L160-L166

**Review by mlissner — CHANGES_REQUESTED** [review]
A lot of comments and suggestions, sorry. But my review is finally done. Two more little ones for HTMX:

1. Does it make sense to put HTMX templates into a subdirectory of `includes`? 
2. No tests for HTMX? I'm surprised. :)

**Comment by mlissner** [comment]
The figma design for logs looks great. No further comments, but let's get an issue going for that if we don't have one already, please. Thank you!

File: cl/alerts/tasks.py (hunk 10 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Done, I've chanted it to `= None` and applied the suggestion:
```
if recap_email_recipients is None:
    recap_email_recipients = []
```
Thanks.

File: cl/alerts/tasks.py (hunk 11 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: I imagined I should have to use something like this, but I wasn't sure what kind of data structure was appropriate so I ended up using that ugly list of tuples. The new data class `DocketAlertRecipient` helps simplify all the things here and is clearer.

I just keep the `recap_email_user_does_not_exist_list` because is just a list of emails so it doesn't match with the new data class.

Thank you!

File: cl/alerts/tasks.py (hunk 12 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Perfect, I added it here and in a different query below where it's also needed to access a related object field.

File: cl/alerts/tasks.py (hunk 13 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Yeah, this simplification was possible thanks to the new data class.

File: cl/alerts/tasks.py (hunk 14 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Yeah, this was removed!

File: cl/alerts/tasks.py (hunk 15 above)
Comment (albertisfu) [reply cl/alerts/tasks.py:130]: Changed!

File: cl/alerts/tasks.py (hunk 17 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Changed, now it's not necessary to pass the recipients to `send_docket_alert_webhooks`.

About the question:
`Why do we need to send alerts before we can send webhooks?`

We don't need to send alerts before we send webhooks but we need to get the recipients of docket alerts before sending the webhooks because when getting the recipients we also create docket alerts for recap email users with auto-subscribe turned on, so we also include those docket alert subscriptions (first-time notifications for a case) when sending webhooks.

File: cl/alerts/tasks.py (hunk 18 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Yeah, it was possible to get the recipients within `send_docket_alert_webhooks`

First, get a list of recipients subscribed to the docket.
```
webhook_recipients = User.objects.filter(
            docket_alerts__docket_id=d_pk,
            docket_alerts__alert_type=DocketAlert.SUBSCRIPTION,
    ).distinct()```
    
Then select the DocketAlert Webhooks from these users.
```
webhooks = Webhook.objects.filter(
        event_type=WebhookEventType.DOCKET_ALERT,
        user__in=webhook_recipients,
        enabled=True,
    )
```

File: cl/alerts/tasks.py (hunk 19 above)
Comment (albertisfu) [reply cl/alerts/tasks.py:316]: I've added it. Now, is set `None`.

File: cl/alerts/tasks.py (hunk 20 above)
Comment (albertisfu) [reply cl/alerts/tasks.py:338]: Well, this `failure_count` belongs to the Webhook object, not the WebhookEvent. That's why is not passed to the `create`.

**Comment by albertisfu** [comment]
Thank you for all your suggestions, they were so useful. 
I hope that with this refactor the code is simpler.

In addition to some comments above. I added the `htmx` templates within subdirectory of `includes`.

Added tests for HTMX endpoints.

I've opened an issue for logs and another for the webhook test feature for users.

File: cl/alerts/tasks.py [hunk 22]
```diff
@@ -168,25 +195,26 @@ def send_docket_alert(
         delete_redis_semaphore("ALERTS", make_alert_key(d_pk))
         return
 
-    # Notify every user that's subscribed to this alert.
-    messages = make_alert_messages(
-        d, new_des, email_addresses, recap_email_recipients
-    )
+    docket_alert_messages = make_alert_messages(d, new_des, da_recipients)
+    messages = docket_alert_messages
```
Comment (mlissner) [inline cl/alerts/tasks.py:199]: What's this about?

File: cl/alerts/tasks.py (hunk 20 above)
Comment (mlissner) [reply cl/alerts/tasks.py:338]: Got it.

File: cl/users/urls.py (hunk 1 above)
Comment (mlissner) [reply cl/users/urls.py]: I think this URL tweak is still needed.

File: cl/alerts/tasks.py (hunk 9 above)
Comment (mlissner) [reply cl/alerts/tasks.py]: Refactor looks good. I still wonder if it could be simplified more, but I'm happy with it. I think it's a big improvement. Hopefully it still works properly. That's a risk, but it's worth taking when the code gets so complex. Thanks for sticking with it through two rounds of review.

File: cl/alerts/tasks.py (hunk 11 above)
Comment (mlissner) [reply cl/alerts/tasks.py]: Looks good. Thanks again.

File: cl/alerts/tasks.py (hunk 18 above)
Comment (mlissner) [reply cl/alerts/tasks.py]: I still think it's possible to do it in one query, but I won't ask you to try. Two is great (and actually pretty clear). :)

**Comment by mlissner** [comment]
One or two things left, but easy stuff. Back to you, thank you.

File: cl/users/urls.py (hunk 1 above)
Comment (albertisfu) [reply cl/users/urls.py]: Sorry, I missed this comment. I've changed it to: `/api/htmx/webhooks/ ` I think it describes pretty well that it belongs to htmx actions.

I moved it into users.urls

File: cl/alerts/tasks.py (hunk 21 above)
Comment (albertisfu) [reply cl/alerts/tasks.py:307]: I think I forget to comment on this:
Interesting. Yeah, I think using the `JSONRenderer()` we ensure the JSON we sent to Webhooks meets the same characteristics we have in API responses. So I changed to use `JSONRenderer()`. Before we were using the JSON encoding included on `requests` where you just needed to pass a `dict`.

File: cl/alerts/tasks.py (hunk 22 above)
Comment (albertisfu) [reply cl/alerts/tasks.py]: Oh yeah, that's not needed, I've removed it.

**Comment by albertisfu** [comment]
I've applied these small changes.
Let me know if there is something else.
Thanks again!

**Review by mlissner — APPROVED** [review]
